### PR TITLE
[wip] Acquisition : Cloudwatch support

### PR DESCRIPTION
**Pull request by buixor**
Initial cloudwatch support, allows to watch a cloudwatch group for streams, based on the last event time.

 - [x] Handle one-shot mode : support dsn, implement OneShot function
 - [x] Documentation
 - [x] Test strategy : ~90% coverage, using the awesome localstack project

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -8,6 +8,7 @@ require (
 	github.com/Microsoft/go-winio v0.4.16 // indirect
 	github.com/antonmedv/expr v1.8.9
 	github.com/appleboy/gin-jwt/v2 v2.6.4
+	github.com/aws/aws-sdk-go v1.34.28
```
Comment (registergoofy) [inline go.mod:11]: It seems that a more recent version exists. Is this intended ?

File: pkg/acquisition/acquisition.go [hunk 2]
```diff
@@ -242,14 +247,17 @@ func StartAcquisition(sources []DataSource, output chan types.Event, AcquisTomb
...
-				return err
+				//if one of the acqusition returns an error, we kill the others to properly shutdown
+				AcquisTomb.Kill(err)
 			}
 			return nil
 		})
 		//register acquisition specific metrics
 		prometheus.MustRegister(subsrc.GetMetrics()...)
 	}
+	log.Infof("waiting on tomb...")
```
Comment (registergoofy) [inline pkg/acquisition/acquisition.go:258]: I would use log sth like "waiting on acquisition tomb" or maybe "waiting for acquisition to terminate"

File: pkg/acquisition/modules/cloudwatch/cloudwatch.go [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+}
+
+var def_DescribeLogStreamsLimit = int64(50)
+var def_PollNewStreamInterval = 10 * time.Second
+var def_MaxStreamAge = 5 * time.Minute
+var def_PollStreamInterval = 10 * time.Second
+var def_AwsApiCallTimeout = 10 * time.Second
+var def_StreamReadTimeout = 10 * time.Minute
+var def_GetLogEventsPagesLimit = int64(1000)
+
```
Comment (registergoofy) [inline pkg/acquisition/modules/cloudwatch/cloudwatch.go:69]: I may be very picky but I would prefer the syntax:
```
var (
   def_DescribeLogStreamsLimit = int64(50)
   def_PollNewStreamInterval = 10 * time.Second
   def_MaxStreamAge = 5 * time.Minute
   def_PollStreamInterval = 10 * time.Second
￼   def_AwsApiCallTimeout = 10 * time.Second
   def_StreamReadTimeout = 10 * time.Minute
   def_GetLogEventsPagesLimit = int64(1000)
)
```

File: pkg/acquisition/modules/cloudwatch/cloudwatch.go [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+									PrependCloudwatchTimestamp: cw.Config.PrependCloudwatchTimestamp,
+								}
+								out <- monitorStream
+							}
+						}
+						if lastPage {
+							cw.logger.Tracef("reached last page")
+							hasMoreStreams = false
+						}
+						return true
```
Comment (AlteredCoder) [inline pkg/acquisition/modules/cloudwatch/cloudwatch.go:229]: is "true" considered as an error ?

File: pkg/acquisition/modules/cloudwatch/cloudwatch.go [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+							startFrom = page.NextToken
+							//we check if the stream has been written to recently enough to be monitored
+							if event.LastIngestionTime != nil {
+								//aws uses millisecond since the epoch
+								oldest := time.Now().UTC().Add(-*cw.Config.MaxStreamAge)
+								LastIngestionTime := time.Unix(0, *event.LastIngestionTime*int64(time.Millisecond))
+								if LastIngestionTime.Before(oldest) {
+									cw.logger.Tracef("stop iteration, %s reached oldest age, stop (%s < %s)", *event.LogStreamName, LastIngestionTime, time.Now().Add(-*cw.Config.MaxStreamAge))
+									hasMoreStreams = false
+									return false
```
Comment (AlteredCoder) [inline pkg/acquisition/modules/cloudwatch/cloudwatch.go:210]: same here (prototype return an error)

**Review by AlteredCoder — APPROVED** [review]

File: pkg/acquisition/modules/cloudwatch/cloudwatch.go (hunk 5 above)
Comment (buixor) [reply pkg/acquisition/modules/cloudwatch/cloudwatch.go:212]: the prototype is `func(page *cloudwatchlogs.DescribeLogStreamsOutput, lastPage bool) bool` -> we return false if we want to stop iteration.

File: pkg/acquisition/modules/cloudwatch/cloudwatch.go (hunk 4 above)
Comment (buixor) [reply pkg/acquisition/modules/cloudwatch/cloudwatch.go:231]: the prototype is `func(page *cloudwatchlogs.DescribeLogStreamsOutput, lastPage bool) bool` -> we return false if we want to stop iteration.

File: go.mod (hunk 1 above)
Comment (buixor) [reply go.mod]: even more, we have the topic of migrating to aws sdk v2, but it's go >= 1.15

File: pkg/acquisition/modules/cloudwatch/cloudwatch.go [hunk 6]
```diff
@@ -0,0 +1,629 @@
...
+	err := yaml.UnmarshalStrict(cfg, &cwConfig)
+	if err != nil {
+		return errors.Wrap(err, "Cannot parse CloudwatchSource configuration")
+	}
+	/*
+		TBD: Acquisition returned error : while describing group acquis-cloudwatch-tests: InvalidParameterException: Cannot order by LastEventTime with a logStreamNamePrefix.
+		either we sort by time or by prefix, can't do both :()
+	*/
+	cw.Config = cwConfig
+	cw.logger = logger.WithField("group", cw.Config.GroupName)
```
Comment (blotus) [inline pkg/acquisition/modules/cloudwatch/cloudwatch.go:94]: Maybe customize the logger after checking GroupName is not empty ?

File: pkg/acquisition/modules/cloudwatch/cloudwatch.go [hunk 7]
```diff
@@ -0,0 +1,629 @@
...
+
+			if cw.Config.StreamName != nil && newStream.StreamName != *cw.Config.StreamName {
+				cw.logger.Tracef("stream %s != %s", newStream.StreamName, *cw.Config.StreamName)
+				continue
+			}
+
+			if cw.Config.StreamRegexp != nil {
+				match, err := regexp.Match(newStream.StreamName, []byte(*cw.Config.StreamRegexp))
+				if err != nil {
+					cw.logger.Warningf("invalid regexp : %s", err)
```
Comment (blotus) [inline pkg/acquisition/modules/cloudwatch/cloudwatch.go:287]: Check if the regexp is valid in the `Configure` method ?

**Pull request merged**